README.md
# reversi

The reversi game made all in clojure

## Usage

1.- lein run

2.- the game start asking who will start the game? only put -1 for black or 1 for white and 
    press enter.

3.- white or black start, white is visualized by + and black for - put the number wher you want
    add a piece.

4.- is the turn for the second player do the same put the number wher you want add your piece

5.- the game is over when there aren't available places

6.- the game is over and you get the result for the game

## tests 

the game works and you can see the test until a complete movement is posible

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-06-30
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-06-30
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/reversi/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/reversi/compare/0.1.0...0.1.1


test/reversi/core_test.clj
(ns reversi.core-test
  (:require [clojure.test :refer :all]
            [reversi.core :as reversi.core]))


(deftest a-test
  (is (= 0 0)))




test/reversi/board_test.clj
(ns reversi.board-test
  (:require [clojure.test :refer :all]
            [reversi.board :as reversi.board]))


(deftest a-test
  (is (= 0 0)))

(deftest connect-left-right-test
  (is (= [61 62 63 64] (reversi.board/connect-left-right 60)))
  (is (= [2 3 4 5 6 7 8] (reversi.board/connect-left-right 1))))

(deftest connect-right-left-test
  (is (= []  (reversi.board/connect-right-left 1)))
  (is (= [22 21 20 19 18 17] (reversi.board/connect-right-left 23))))

(deftest connect-up-down-test
  (is (= [29 37 45 53 61] (reversi.board/connect-up-down 21)))
  (is (= [9 17 25 33 41 49 57] (reversi.board/connect-up-down 1))))

(deftest connect-down-up-test
  (is (= [32 24 16 8] (reversi.board/connect-down-up 40)))
  (is (= [55 47 39 31 23 15 7] (reversi.board/connect-down-up 63))))

(deftest connect-left-down-test
  (is (= [12 19 26 33] (reversi.board/connect-left-down 5)))
  (is (= [27 34 41] (reversi.board/connect-left-down 20))))

(deftest connect-right-down-test
  (is (= [10 19 28 37 46 55 64] (reversi.board/connect-right-down 1)))
  (is (= [32] (reversi.board/connect-right-down 23))))

(deftest connect-right-up-test
  (is (= [43 36 29 22 15 8] (reversi.board/connect-right-up 50)))
  (is (= [52 45 38 31 24] (reversi.board/connect-right-up 59))))

(deftest connect-left-up-test
  (is (= [55 46 37 28 19 10 1] (reversi.board/connect-left-up 64)))
  (is (= [] (reversi.board/connect-left-up 5))))

(deftest complete-board-test
  (is (= {1 {:left-down [],
             :left-right [2 3 4 5 6 7 8],
             :right-down [10 19 28 37 46 55 64],
             :right-up [],
             :left-up [],
             :up-down [9 17 25 33 41 49 57],
             :right-left [],
             :down-up [],
             :occupied-by 0},
          2 {:left-down [9],
             :left-right [3 4 5 6 7 8],
             :right-down [11 20 29 38 47 56],
             :right-up [],
             :left-up [],
             :up-down [10 18 26 34 42 50 58],
             :right-left [1],
             :down-up [],
             :occupied-by 0},
          3 {:left-down [10 17],
             :left-right [4 5 6 7 8],
             :right-down [12 21 30 39 48],
             :right-up [],
             :left-up [],
             :up-down [11 19 27 35 43 51 59],
             :right-left [2 1],
             :down-up [],
             :occupied-by 0},
          4 {:left-down [11 18 25],
             :left-right [5 6 7 8],
             :right-down [13 22 31 40],
             :right-up [],
             :left-up [],
             :up-down [12 20 28 36 44 52 60],
             :right-left [3 2 1],
             :down-up [],
             :occupied-by 0},
          5 {:left-down [12 19 26 33],
             :left-right [6 7 8],
             :right-down [14 23 32],
             :right-up [],
             :left-up [],
             :up-down [13 21 29 37 45 53 61],
             :right-left [4 3 2 1],
             :down-up [],
             :occupied-by 0}} (reversi.board/complete-board 5))))


src/reversi/io.clj
(ns reversi.io
  (:require [reversi.board :as reversi.board]
            [reversi.move :as reversi.move]))

(declare black-turn)


(defn how-many-pieces
  "count how many pieces of a color there are in board"
  [board color]
  (count (filter #(= % color) (map #(get-in board [% :occupied-by]) (range 1 65)))))

(defn game-over
  "Announce the game is over and who is the winner"
  [board]
  (let [total-white (how-many-pieces board 1)
        total-black (how-many-pieces board -1)
        winner (if (> (how-many-pieces board 1) (reversi.io/how-many-pieces board -1))
                 "white"
                 "black")]
    (println "the black had: " total-black "the white had " total-white "; the winner is" winner )))

(defn white-turn
  [board]
  (println (reversi.board/print-board board))
  (println "choose the number where you want put a piece you will be + for white")
  (let [pos (read)]
    (if (reversi.move/valide-move? board pos)
     (let [new-board (reversi.move/complete-white-movement board pos)]
    (if (> (how-many-pieces new-board 0) 0)
    (black-turn new-board)
    (game-over new-board)))
     (do (println "ingresaste una casilla invalida, escoge otra")
         (white-turn board)))))

(defn black-turn
  [board]
  (println (reversi.board/print-board board))
  (println "choose the number where you want put a piece you will be - for black")
  (let [pos (read)]
    (if (reversi.move/valide-move? board pos)
      (let [new-board (reversi.move/complete-black-movement board pos)]
        (if (> (how-many-pieces new-board 0) 0)
          (white-turn new-board)
          (game-over new-board)))
      (do (println "ingresaste una casilla invalida, escoge otra")
          (black-turn board)))))








src/reversi/move.clj
(ns reversi.move
  (:require [reversi.board :as reversi.board]))

(defn occupied-by?
  "occupied by whom is the position"
  [board pos]
  (get-in board [pos :occupied-by]))

(defn add-a-white
  "put a white (1) piece at given position in the board"
  [board pos]
  (assoc-in board [pos :occupied-by] 1))

(defn add-a-black
  "put a black (-1) piece at given position in the board"
  [board pos]
  (assoc-in board [pos :occupied-by] -1))

(defn valide-move?
  "the only restriction for (for the moment) a valid move is a empty place"
  [board pos]
  (= (get-in board [pos :occupied-by]) 0))

(defn occupied?
  [board pos connection]
  "it gives a vector with color in the left-right connections of a position "
  (vec (map #(get-in board [% :occupied-by])
            (get-in board [pos connection]))))

(defn occupied-left-right?
  "it gives a vector with color in the left-right connections of a position "
  [board pos]
  (occupied? board pos :left-right))

(defn occupied-right-left?
  "it gives a vector with color in the right-left connections of a position "
  [board pos]
  (occupied? board pos :right-left))

(defn occupied-left-up?
  "it gives a vector with color in the left-up connections of a position "
  [board pos]
  (occupied? board pos :left-up))

(defn occupied-left-down?
  "it gives a vector with color in the left-down connections of a position "
  [board pos]
  (occupied? board pos :left-down))

(defn occupied-right-up?
  "it gives a vector with color in the right-up connections of a position "
  [board pos]
  (occupied? board pos :right-up))

(defn occupied-right-down?
  "it gives a vector with color in the right-down connections of a position "
  [board pos]
  (occupied? board pos :right-down))

(defn occupied-up-down?
  "it gives a vector with color in the up-down connections of a position "
  [board pos]
  (occupied? board pos :up-down))

(defn occupied-down-up?
  "it gives a vector with color in the down-up connections of a position "
  [board pos]
  (occupied? board pos :down-up))

(defn other-equal?
  "check if there is other with the same color in a conection"
  [board pos color connection]
  (some #(= % color) (connection board pos)))

(defn take-the-center
  "a vector with the information occupied-by between 2 pieces with the same color"
  [board pos color connection]
  (vec (take-while #(not= % color) (connection board pos))))

(defn all-the-center?
  "check if all the center between two pieces with the same color are all of the opposite color"
  [board pos color connection]
  (if (other-equal? board pos color connection)
    (= (vec (take (count (vec (take-while #(not= % color) (connection board pos)))) (repeat (* -1 color))))
       (vec (take-while #(not= % color) (connection board pos))))
    false))

(defn changes
  "make all the changes for a connection"
  [board pos occupied connection add color]
  (let [board-after-add (add board pos)]
    (if (all-the-center? board pos color occupied)
      (reduce (fn [board pos] (add board pos))
              board-after-add
              (take (count (take-the-center board pos color occupied)) (connection pos)))
      board-after-add)))

(defn changes-black-right-left
  "make all the changes in right left connection if exist (after a black turn)"
  [board pos]
  (changes board pos occupied-right-left? reversi.board/connect-right-left add-a-black -1))

(defn changes-black-left-right
  "make all the changes in left-right connection if exist (after a black  turn)"
  [board pos]
  (changes board pos occupied-left-right? reversi.board/connect-left-right add-a-black -1))

(defn changes-black-up-down
  "make all the changes in up-down connection if exist (after a black  turn)"
  [board pos]
  (changes board pos occupied-up-down? reversi.board/connect-up-down add-a-black -1))

(defn changes-black-down-up
  "make all the changes in down-up connection if exist (after a black  turn)"
  [board pos]
  (changes board pos occupied-down-up? reversi.board/connect-down-up add-a-black -1))

(defn changes-black-left-up
  "make all the changes in left-up connection if exist (after a black  turn)"
  [board pos]
  (changes board pos occupied-left-up? reversi.board/connect-left-up add-a-black -1))

(defn changes-black-left-down
  "make all the changes in left-down connection if exist (after a black  turn)"
  [board pos]
  (changes board pos occupied-left-down? reversi.board/connect-left-down add-a-black -1))

(defn changes-black-right-down
  "make all the changes in right-down connection if exist (after a black  turn)"
  [board pos]
  (changes board pos occupied-right-down? reversi.board/connect-right-down add-a-black -1))

(defn changes-black-right-up
  "make all the changes in right-up connection if exist (after a black  turn)"
  [board pos]
  (changes board pos occupied-right-up? reversi.board/connect-right-up add-a-black -1))

(defn complete-black-movement
  "make all the changes after add a black piece"
  [board pos]
  (if (valide-move? board pos)
    (reduce (fn [new-board changes-conection-fn] (changes-conection-fn new-board pos))
            board
            [changes-black-right-up
             changes-black-right-down
             changes-black-left-down
             changes-black-left-up
             changes-black-down-up
             changes-black-up-down
             changes-black-left-right
             changes-black-right-left])))

(defn changes-white-right-left
  "make all the changes in right left connection if exist (after a white turn)"
  [board pos]
  (changes board pos occupied-right-left? reversi.board/connect-right-left add-a-white 1))

(defn changes-white-left-right
  "make all the changes in left-right connection if exist (after a white turn)"
  [board pos]
  (changes board pos occupied-left-right? reversi.board/connect-left-right add-a-white  1))

(defn changes-white-up-down
  "make all the changes in up-down connection if exist (after a white turn)"
  [board pos]
  (changes board pos occupied-up-down? reversi.board/connect-up-down add-a-white 1))

(defn changes-white-down-up
  "make all the changes in down-up connection if exist (after a white turn)"
  [board pos]
  (changes board pos occupied-down-up? reversi.board/connect-down-up add-a-white 1))

(defn changes-white-left-up
  "make all the changes in left-up connection if exist (after a white turn)"
  [board pos]
  (changes board pos occupied-left-up? reversi.board/connect-left-up add-a-white 1))

(defn changes-white-left-down
  "make all the changes in left-down connection if exist (after a white turn)"
  [board pos]
  (changes board pos occupied-left-down? reversi.board/connect-left-down add-a-white 1))

(defn changes-white-right-down
  "make all the changes in right-down connection if exist (after a white turn)"
  [board pos]
  (changes board pos occupied-right-down? reversi.board/connect-right-down add-a-white 1))

(defn changes-white-right-up
  "make all the changes in right-up connection if exist (after a white turn)"
  [board pos]
  (changes board pos occupied-right-up? reversi.board/connect-right-up add-a-white 1))

(defn complete-white-movement
  "make all the changes after add a black piece"
  [board pos]
  (if (valide-move? board pos)
    (reduce (fn [new-board changes-conection-fn] (changes-conection-fn new-board pos))
            board
            [changes-white-right-up
             changes-white-right-down
             changes-white-left-down
             changes-white-left-up
             changes-white-down-up
             changes-white-up-down
             changes-white-left-right
             changes-white-right-left])))




src/reversi/board.clj
(ns reversi.board)

(defn row
  "the row number at the position"
  [position]
  (int (Math/ceil (/ position 8))))

(defn column
  "the column number at the position"
  [position]
  (- position (* (dec (row position)) 8)))

(defn connect-left-right
  "the right left connections"
  [position]
  (vec (range (inc position) (inc (* 8 (row position))))))

(defn connect-right-left
  "the right left connections"
  [position]
  (vec (reverse (range (inc (* 8 (dec (row position)))) position))))

(defn connect-up-down
  [position]
  (vec (range (+ position 8) 65 8)))

(defn connect-down-up
  [position]
  (vec (reverse (range (- position (* 8 (dec (row position)))) position 8))))

(defn connect-left-down
  [position]
  (vec (take (dec (column position)) (range (+ position 7) 64 7))))

(defn connect-right-down
  [position]
  (vec (take (- 8 (column position)) (range (+ position 9) 65 9))))

(defn connect-right-up
  [position]
  (vec (take (- 8 (column position)) (range (- position 7) 0 -7))))

(defn connect-left-up
  [position]
  (vec (take (dec (column position)) (range (- position 9) 0 -9))))

(defn connect-left-right-map
  "map with left-right conections"
  [board pos]
  (assoc-in board [pos :left-right] (connect-left-right pos)))

(defn connect-right-left-map
  "map with right-left conections"
  [board pos]
  (assoc-in board [pos :right-left] (connect-right-left pos)))

(defn connect-up-down-map
  "map with up-down conections"
  [board pos]
  (assoc-in board [pos :up-down] (connect-up-down pos)))

(defn connect-down-up-map
  "map with down-up conections"
  [board pos]
  (assoc-in board [pos :down-up] (connect-down-up pos)))

(defn connect-left-down-map
  "map with left-down conections"
  [board pos]
  (assoc-in board [pos :left-down] (connect-left-down pos)))

(defn connect-left-up-map
  "map with left-up conections"
  [board pos]
  (assoc-in board [pos :left-up] (connect-left-up pos)))

(defn connect-right-up-map
  "map with right-up conections"
  [board pos]
  (assoc-in board [pos :right-up] (connect-right-up pos)))

(defn connect-right-down-map
  "map with right-down conections"
  [board pos]
  (assoc-in board [pos :right-down] (connect-right-down pos)))

(defn all-conections-number
  "create all conections for a number"
  [board1 positions]
  (let [occupied-board (assoc-in board1 [positions :occupied-by] 0)]
    (reduce (fn [board all-connections]
              (all-connections board positions))
            occupied-board
            [connect-right-left-map connect-left-right-map connect-up-down-map connect-down-up-map connect-left-down-map connect-left-up-map connect-right-down-map connect-right-up-map])))

(defn complete-board
  "create a complete board"
  [positions]
  (reduce (fn [board positions]
            (all-conections-number board positions))
          {}
          (range 1 (inc positions))))

(defn occupied-representation
  [board pos]
  (cond (= (get-in board [pos :occupied-by]) 0) "n"
        (= (get-in board [pos :occupied-by]) 1) "+"
        (= (get-in board [pos :occupied-by]) -1) "-"))

(defn print-board
  "print a board I know, I need modify the function but this gives a beautiful board"
  [board]
  (do (println (vec (map #(str "  " % (occupied-representation board %))
                (range 1 9)) ))
      (println (into [(str "  " 9 (occupied-representation board 9))] (vec (map #(str " " % (occupied-representation board %))
                    (range 10 17)))))
      (println (vec (map #(str  " " % (occupied-representation board %))
                    (range 17 25))) )
      (println (vec (map #(str " " % (occupied-representation board %))
                    (range 25 33))) )
      (println (vec (map #(str " " % (occupied-representation board %))
                    (range 33 41))) )
      (println (vec (map #(str " " % (occupied-representation board %))
                    (range 41 49))) )
      (println (vec (map #(str " " % (occupied-representation board %))
                    (range 49 57))) )
      (print (vec (map #(str " " % (occupied-representation board %))
                    (range 57 65))))))


